```{r}
setwd("~/Documents/Internship/InfoGrab")

required_packages <- c(
  "shiny", "shinyBS", "shinythemes", "shinycustomloader", "shinycssloaders", 
  "ggplot2", "pheatmap", "dplyr", "readr", "readxl", 
  "pheatmap", "DT", "stringr", "purrr", "tibble", "reshape2", 
  "RColorBrewer", "gplots", "scales", "data.table", "gridExtra", 
  "plotly", "lubridate", "shinyjs", "shinydashboard", 
  "shinycssloaders", "shinyWidgets", "htmltools", "tools", 
  "htmlwidgets", "DESeq2", "ggplotify", "grid", "graphics", 
  "igraph", "seriation", "GGally", "knitr", "devtools", "stringr",
  "JBrowseR"
)

install_if_missing <- function(pkg) {
  if (!require(pkg, character.only = TRUE)) {
    install.packages(pkg, dependencies = TRUE)
  }
}

lapply(required_packages, install_if_missing)

if (!require(DESeq2)) {
  BiocManager::install("DESeq2", force = TRUE)
}

if (!require(pcaExplorer)) {
  BiocManager::install("pcaExplorer", force = TRUE)
}

if (!require(tispec)) {
  library(devtools)
  install_github('roonysgalbi/tispec')
}

lapply(required_packages, library, character.only = TRUE)

library(shiny)
library(shinyBS)
library(shinythemes)
library(shinyjs)
library(shinydashboard)
library(shinycssloaders)
library(shinycustomloader)
library(dplyr)
library(ggplot2)
library(ggrepel)
library(pheatmap)
library(DT)
library(stringr)
library(viridis)
library(readr)
library(tidyr)
library(tibble)
library(purrr)
library(magrittr)
library(htmltools)
library(scales)
library(plotly)
library(DESeq2)
library(pcaExplorer)
library(ggplotify)
library(htmlwidgets)
library(tools)
library(grid)
library(graphics)
library(igraph)
library(GGally)
library(gridExtra)
library(tispec)
library(knitr)
library(stringr)

if (!exists("comparison_results")) {
  comparison_results <- readRDS("data/comparison_results.rds")
}

if (!exists("RPKM_data")) {
  RPKM_data <- read.delim("data/logRPKM0625_filter.txt")
}

if (!exists("phenodata")) {
  phenodata <- read.csv("data/Pheno_data071523.csv")
}

if (!exists("cnt_data")) {
  cnt_data <- read.delim("data/all_cnt_0625.txt")
}

mycolors1 <- c(
  `Kidney`="#43009A",
  `Trachea`="#990099",
  `BCell`="#0DFAFA",
  `TCell(spleen)`="#13B7B7",
  `Bursa`="#004C99",
  `Thymus`="#2685E4",
  `Macrophage(D0)`="#F02B6D",
  `Macrophage(D3)`="#FF0091",
  `Macrophage(D6)`="#F572BC",
  `Macrophage(lung)`="#D06AAA",
  `Monocyte(blood)`="#91155B",
  `Ileum`="#98D55A",
  `Jejunum`="#4C9900",
  `Proximal.Cecum`="#CCFF99",
  `Iliotibial.major`="#A1122A",
  `Pectoralis.major`="#FFCCCC",
  `Isthmus`="#FF8000",
  `Magnum`="#DE5100",
  `Shell.Gland`="#FFC78E",
  `Ovary`="#CCCC00"
)

mycolors2 <- c(
  `Immune` = "#31E1F7", 
  `Respiratory` = "#400D51", 
  `Excretory` ="#FF7777",
  `Muscular` = "#D800A6",
  `Intestine` ="#6499E9", 
  `Reproductive` = "#836FFF")

```

```{r}

############################################################

# INFO GRAB

# Cite BioRender
# Cite pcaExplorer

# Need to update how the data loads

tissue_names <- unique(phenodata$Tissue)

systems <- unique(phenodata$System)
tissue_by_system <- lapply(systems, function(system) {
  unique(phenodata$Tissue[phenodata$System == system])
})
names(tissue_by_system) <- systems

##################

ui <- navbarPage(
  title = tags$a(
    tags$div(
      tags$a(
        href = "https://www.washington.edu",
        target = "_blank",
        tags$img(src = "UW_logo.png", height = "50px", style = "padding: 5px; margin-top: -17px;")
      ),
      tags$a(
        href = "https://www.westernu.edu",
        target = "_blank",
        tags$img(src = "WesternU_logo.png", height = "60px", style = "padding: 5px; margin-top: -19px;")
      )
      
    )
  ),
  
  windowTitle = "INFO GRAB",
  theme = shinytheme("paper"),
  id = "navbar",
  useShinyjs(),
  tags$head(
    tags$link(rel = "stylesheet", href = "https://fonts.googleapis.com/css2?family=Sen:wght@400;700&display=swap"),
    tags$link(rel = "stylesheet", href = "https://fonts.googleapis.com/css2?family=Ubuntu:wght@400;700&display=swap"),
    tags$script(HTML("
      $(document).on('shiny:connected', function(event) {
        Shiny.setInputValue('startAnalysis', true);
      });
    ")),
    tags$style(HTML("
      .navbar-nav > li > a, .navbar-brand {
        font-family: 'Sen', sans-serif !important;
        font-size: 15px;
      }
      h1 {
        font-family: 'Ubuntu', sans-serif !important;
        font-size: 40px !important;
        font-weight: bold !important;
      }
      h2 {
        font-family: 'Ubuntu', sans-serif !important;
        font-size: 34px !important;
        font-weight: bold !important;
      }
      h3 {
        font-family: 'Ubuntu', sans-serif !important;
        font-size: 28px !important;
        font-weight: bold !important;
      }
      h4 {
        font-family: 'Sen', sans-serif !important;
        font-size: 20px !important;
      }
      .form-group > label {
        font-family: 'Sen', sans-serif !important;
      }
      .navbar-nav > li > a, .navbar-brand {
        font-family: 'Sen', sans-serif !important;
        font-size: 15px;
        color: white !important;
      }
      .navbar-default {
        background-color: #4B0082 !important;
        border-color: #4B0082 !important;
      }
      .navbar-default .navbar-nav > .active > a, 
      .navbar-default .navbar-nav > .active > a:focus, 
      .navbar-default .navbar-nav > .active > a:hover {
        color: white !important;
        background-color: grey !important;
      }
      .btn {
        background-color: #4B0082 !important;
        color: white !important;
      }
      .btn:hover {
        background-color: #4B0082 !important;
        color: white !important;
      }
      #info-grab-link:hover {
        background-color: grey !important;
        color: white !important;
      }
      #progress-container {
        position: fixed;
        bottom: 10px;
        right: 10px;
        width: 300px;
        z-index: 1050;
      }
    "))
  ),
  tabPanel("INFO GRAB",
           fluidPage(
             div(style = "text-align: center;",
                 h2("INFO GRAB"),
                 h3("INteractive Functional Ontology Genomic Regulatory Element Analysis and Browsing"),
                 tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                 h5("Welcome to INFO GRAB. This app allows you to perform differential expression analysis and 
                   visualize the results through various plots and tables, looking at tissues in the chicken (Gallus gallus)."),
                 tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                 img(src = "chicken.png", id = "chicken-image", style = "width: 40%; max-width: 100%; height: auto;")
             ),
             uiOutput("tissue_display")
           )
  ),
  
  tabPanel("Differential Analysis",
           tabsetPanel(
             id = "differential_analysis_tab",
             tabPanel("Home",
                      sidebarLayout(
                        sidebarPanel(
                          selectInput("tissue_select1", "Tissue 1", choices = tissue_names),
                          selectInput("tissue_select2", "Tissue 2", choices = tissue_names[-1]),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          actionButton("add_all_de_cart", "Add All DE Genes to Gene Cart"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          selectInput("selected_fdr", "FDR", 
                                      choices = c("No filter" = NA, "0.01" = 0.01, "0.05" = 0.05, "0.1" = 0.1), 
                                      selected = "0.05"),
                          sliderInput("selected_fc", "FC (log2 fold change)", min = 0, max = 20, step = 1, value = 0),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          sliderInput("num_of_digits", "Number of digits", value = 5, min = 2, max = 20),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          downloadButton("download_DEA_data", "Download CSV"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          tags$p("FDR: The expected proportion of false discoveries among the rejected hypotheses."),
                          tags$p("FC: A measure describing how much a quantity changes between an original and a subsequent measurement."),
                          uiOutput("fold_change_equation")
                        ),
                        mainPanel(
                          uiOutput("title"),
                          
                          withLoader(tableOutput("summary_info"), type = "html", loader = "dnaspin"),
                          br(),
                          textInput("searched_gene", "Search for genes (comma-separated)", width = '600px'),
                          actionButton("search_gene", "Search"),
                          actionButton("random_gene", "Search 10 Random Genes"),
                          actionButton("clear_search_main", "Clear"),
                          actionButton("add_to_cart_home", "Add to Gene Cart"),
                          DTOutput("searched_gene_info"),
                          br(),
                          uiOutput("collapsePanels")
                          
                        )
                      )
             ),
             
             tabPanel("Correlation Plot",
                      sidebarLayout(
                        sidebarPanel(
                          radioButtons("correlation_mode", "Select Mode:",
                                       choices = c("Selected Tissues" = "selected", "Tissues by System" = "system"),
                                       selected = "selected"),
                          conditionalPanel(
                            condition = "input.correlation_mode == 'selected'",
                            checkboxInput("use_filtered_data", "Use Filtered Data", value = TRUE)
                          ),
                          conditionalPanel(
                            condition = "input.correlation_mode == 'system'",
                            selectInput("correlation_system", "Select System:", choices = systems)
                          ),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          downloadButton("download_correlation_plot", "Download PNG"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          tags$p("This section allows you to visualize the correlation between the expression levels of selected genes across different tissues or systems. Correlation plots are useful for understanding how gene expression patterns are related between different conditions or tissue types."),
                          tags$p("The diagonal panels show density plots of the gene expression distribution for each sample, while the upper panels display the correlation coefficients between pairs of samples.")
                        ),
                        mainPanel(
                          uiOutput("title_cor_plot"),
                          withLoader(plotOutput("correlation_plot", width = "90%", height = "750px"), type = "html", loader = "dnaspin")
                        )
                      )
             ),
             
             tabPanel("Volcano Plot",
                      sidebarLayout(
                        sidebarPanel(
                          textInput("searched_gene_volcano", "Search for gene:", width = '600px'),
                          actionButton("search_gene_volcano", "Search"),
                          actionButton("add_to_cart_volcano", "Add to Gene Cart"),
                          actionButton("clear_search_gene_volcano", "Clear"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          
                          radioButtons("labeling_option", "Labeling Options:",
                                       choices = list("Basic labels" = "basic", "Advanced labels" = "advanced"),
                                       selected = "basic"),
                          
                          conditionalPanel(
                            condition = "input.labeling_option == 'basic'",
                            sliderInput("n_labels", "Number of Labels", min = 0, max = 100, value = 10)
                          ),
                          
                          conditionalPanel(
                            condition = "input.labeling_option == 'advanced'",
                            sliderInput("n_labels_left", "Number of Labels (Left)", min = 0, max = 50, value = 5),
                            sliderInput("n_labels_right", "Number of Labels (Right)", min = 0, max = 50, value = 5)
                          ),
                          
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          selectInput("volcano_background_col", "Background", 
                                      choices = c("Light" = "White","Dark" = "Grey"), selected = "White"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          downloadButton("download_volcano_plot", "Download PNG"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          tags$p("Each point represents a gene, plotted by its log2 fold change and the negative 
                          logarithm of its adjusted p-value. Genes with significant changes (high fold change 
                                 and low p-value) are highlighted."),
                          tags$p("Genes with the largest log2 fold changes (farthest from the center on the x-axis) 
                                 show the most substantial differences in expression between the two tissues. 
                                 The y-axis represents the significance level, with higher points indicating more 
                                 statistically significant changes.")
                        ),
                        mainPanel(
                          uiOutput("title_volcano_plot"),
                          withLoader(plotOutput("volcano_plot", width = "90%", height = "750px"), type = "html", loader = "dnaspin")
                        )
                      )
             ),
             tabPanel("Heatmap",
                      sidebarLayout(
                        sidebarPanel(
                          selectInput("sort_by", "Sort by:", choices = c("FDR" = "FDR", "FC" = "FC")),
                          selectInput("top_n", "Number of Top Genes:", 
                                      choices = c("5" = 5, "10" = 10, "20" = 20, "50" = 50, "100" = 100, "500" = 500, "1000" = 1000, "All DE Genes" = Inf), 
                                      selected = "10"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          textInput("search_genes_heatmap", "Search for Specific Genes (comma-separated):"),
                          actionButton("search_gene_heatmap", "Search"),
                          actionButton("random_gene_heatmap", "Search 10 Random Genes"),
                          actionButton("clear_search", "Clear"),
                          actionButton("add_to_cart_heatmap", "Add to Gene Cart"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          selectInput("color_scale", "Select Color Scale", 
                                      choices = c("Blue-White-Red" = "blue_white_red", 
                                                  "Green-Black-Red" = "green_black_red",
                                                  "Purple-White-Green" = "purple_white_green",
                                                  "Viridis" = "viridis",
                                                  "Plasma" = "plasma",
                                                  "Cividis" = "cividis",
                                                  "Inferno" = "inferno")),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          downloadButton("download_heatmap_plot", "Download PNG"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          tags$p("Each cell represents the expression level of a gene in a sample, with colors indicating 
                                 the level of expression."),
                          tags$p("The color scale is based on z-scores, which represent the number of standard deviations 
                                 away from the mean expression level. High positive z-scores indicate higher expression levels, 
                                 while high negative z-scores indicate lower expression levels.")
                        ),
                        mainPanel(
                          uiOutput("title_heatmap"),
                          withLoader(plotOutput("heatmap_plot", width = "85%", height = "750px", 
                                                brush = brushOpts(id = "heatmap_brush", resetOnNew = TRUE)), 
                                     type = "html", loader = "dnaspin")
                        )
                      )
             ),
             
             tabPanel("PCA",
                      sidebarLayout(
                        sidebarPanel(
                          radioButtons("pca_mode", "Select Mode:",
                                       choices = c("Selected Tissues" = "selected", "Tissues by System" = "system", "All Tissues" = "all"),
                                       selected = "selected"),
                          conditionalPanel(
                            condition = "input.pca_mode == 'system'",
                            selectInput("pca_system_select", "Select System:", choices = unique(phenodata$System))
                          ),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          downloadButton("download_pca_plot", "Download PNG"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          tags$p("This section allows you to perform Principal Component Analysis (PCA) on your data. PCA is a dimensionality reduction technique that projects high-dimensional data into a lower-dimensional space, making it easier to visualize patterns and relationships in the data."),
                          tags$p("The first two principal components (PC1 and PC2) are plotted, showing the most variance in the data. Each point represents a sample, colored by tissue type and shaped by system.")
                        ),
                        mainPanel(
                          uiOutput("title_pca_plot"),
                          withLoader(plotOutput("pca_plot", height = "600px"), type = "html", loader = "dnaspin")
                        )
                      )
             ),
             
             tabPanel("Gene Expression Chart",
                      sidebarLayout(
                        sidebarPanel(
                          textInput("barplot_searched_gene", "Search for genes (comma-separated)", width = '600px'),
                          actionButton("search_gene_barplot", "Search"),
                          actionButton("random_gene_barplot", "Search 5 Random Genes"),
                          actionButton("clear_search_gene_barplot", "Clear"),
                          actionButton("add_to_cart_barplot", "Add to Gene Cart"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          downloadButton("download_barplot", "Download PNG"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          tags$p("This section provides a barplot of the average expression levels of selected genes across different tissues. This visualization is helpful for comparing how genes are expressed in different tissue types."),
                          tags$p("You can search for specific genes and display their average expression levels across the selected tissues. The plot allows for a quick comparison of gene expression patterns and identification of tissue-specific expression.")
                        ),
                        mainPanel(
                          uiOutput("dynamic_title"),
                          uiOutput("dynamic_barplot_output")
                        )
                      )
             ),
             
             
             tabPanel("Help",
                      mainPanel(
                        h2("Analysis"),
                        
                        h3("Home"),
                        p("In the 'Home' tab, you can select two tissues to compare and adjust the analysis parameters."),
                        tags$ul(
                          tags$li(tags$b("Tissue 1 and Tissue 2:"), " Select the tissues you want to compare."),
                          tags$li(tags$b("FDR:"), " Choose the false discovery rate threshold for filtering the results."),
                          tags$li(tags$b("FC (log2 fold change):"), " Set the fold change threshold for filtering the results."),
                          tags$li(tags$b("Number of digits:"), " Specify the number of decimal places for displaying values."),
                          tags$li(tags$b("Download CSV:"), " Download the filtered data as a CSV file."),
                          tags$li(tags$b("Search for genes:"), " Enter genes to search in the results."),
                          tags$li(tags$b("Search 10 Random Genes:"), " Display information for 10 randomly selected genes."),
                          tags$li(tags$b("Clear:"), " Clear the gene search input."),
                          tags$li(tags$b("Fold Change Equation:"), " Displays the log2 fold change equation for the selected tissues.")
                        ),
                        
                        h3("Correlation Plot"),
                        p("The 'Correlation Plot' tab allows you to visualize the correlation between expression levels of genes across different tissues."),
                        tags$ul(
                          tags$li(tags$b("Select Mode:"), " Choose between 'Selected Tissues' and 'Tissues by System'."),
                          tags$li(tags$b("Use Filtered Data:"), " Option to use filtered data based on selected FDR and FC thresholds."),
                          tags$li(tags$b("Select System:"), " Select a specific system to view correlation within that system."),
                          tags$li(tags$b("Correlation Plot Legend:"), " The plot includes a legend that matches the tissue colors with their corresponding tissues.")
                        ),
                        
                        h3("Volcano Plot"),
                        p("The 'Volcano Plot' tab allows you to visualize the results of the differential expression analysis."),
                        tags$ul(
                          tags$li(tags$b("Search for gene:"), " Enter a gene name to highlight it on the plot."),
                          tags$li(tags$b("Labeling Options:"), " Choose between basic or advanced labeling options for the plot."),
                          tags$li(tags$b("Background:"), " Choose the background color for the plot."),
                          tags$li(tags$b("Download PNG:"), " Download the volcano plot as a PNG file."),
                          tags$li(tags$b("Advanced Labeling Options:"), " Allows the user to specify the number of labels on each side of the volcano plot separately.")
                        ),
                        
                        h3("Heatmap"),
                        p("The 'Heatmap' tab provides a heatmap visualization of the expression levels of genes."),
                        tags$ul(
                          tags$li(tags$b("Sort by:"), " Choose to sort the genes by FDR or fold change."),
                          tags$li(tags$b("Number of Top Genes:"), " Select the number of top genes to include in the heatmap."),
                          tags$li(tags$b("Search for Specific Genes:"), " Enter gene names (comma-separated) to highlight on the heatmap."),
                          tags$li(tags$b("Random Gene Search:"), " Search for 10 random genes and highlight them on the heatmap."),
                          tags$li(tags$b("Select Color Scale:"), " Choose the color scale for the heatmap."),
                          tags$li(tags$b("Download PNG:"), " Download the heatmap as a PNG file.")
                        ),
                        
                        h3("PCA"),
                        p("The 'PCA' tab allows you to perform Principal Component Analysis on the selected tissues."),
                        tags$ul(
                          tags$li(tags$b("Select Mode:"), " Choose between 'Selected Tissues', 'Tissues by System', and 'All Tissues'."),
                          tags$li(tags$b("Select System:"), " Select a specific system for PCA analysis when 'Tissues by System' mode is selected."),
                          tags$li(tags$b("Download PCA Plot:"), " Download the PCA plot as a PNG file.")
                        ),
                        
                        h3("Gene Expression Chart"),
                        p("The 'Gene Expression Chart' tab allows you to visualize the average expression levels of selected genes across different tissues."),
                        tags$ul(
                          tags$li(tags$b("Search for genes:"), " Enter genes (comma-separated) to create bar charts of their average expression levels."),
                          tags$li(tags$b("Search 5 Random Genes:"), " Display bar charts for 5 randomly selected genes."),
                          tags$li(tags$b("Clear:"), " Clear the gene search input."),
                          tags$li(tags$b("Download Barplot:"), " Download the bar charts as a PNG file.")
                        ),
                        
                        h3("Parameters Explanation"),
                        tags$p("False Discovery Rate (FDR): The expected proportion of false discoveries among the rejected hypotheses."),
                        tags$p("Fold Change (FC): A measure describing how much a quantity changes between an original and a subsequent measurement.")
                      )
             )
             
             
           )
  ),
  
  tabPanel("Tissue-Specific Analysis",
           tabsetPanel(
             tabPanel("Heatmap",
                      sidebarLayout(
                        sidebarPanel(
                          selectInput("n_variable_genes", "Number of most variable genes", choices = c(50, 100, 500, 1000), selected = 500),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          checkboxGroupInput("system_choice", "Select System(s):",
                                             choices = c("All Systems", unique(phenodata$System)),
                                             selected = "All Systems"),
                          uiOutput("tissue_selection_ui"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          selectInput("color_scale_tissue_specific", "Select Color Scale", 
                                      choices = c("Purple-White-Green" = "purple_white_green",
                                                  "Blue-White-Red" = "blue_white_red", 
                                                  "Green-Black-Red" = "green_black_red",
                                                  "Blue-Yellow-Purple" = "cyan_yellow_purple",
                                                  "Viridis" = "viridis",
                                                  "Plasma" = "plasma",
                                                  "Cividis" = "cividis",
                                                  "Inferno" = "inferno"),
                                      selected = "purple_white_green"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          actionButton("add_to_cart_tissue_specific_heatmap", "Add to Gene Cart"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          downloadButton("download_tissue_specific_heatmap", "Download PNG"),
                          tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
                          tags$p("This section displays a tissue-specific heatmap, which focuses on the expression of genes that are most variable across the selected tissues."),
                          tags$p("The heatmap is color-coded to represent different levels of gene expression, allowing you to easily identify patterns and differences in gene expression between tissues. This is particularly useful for identifying tissue-specific genes and understanding how gene expression varies across different biological systems.")
                        ),
                        mainPanel(
                          withLoader(plotOutput("tissue_specific_heatmap", width = "100%", height = "750px", 
                                                click = "heatmap_click", brush = brushOpts(id = "heatmap_tissue_specific_brush", resetOnNew = TRUE)), 
                                     type = "html", loader = "dnaspin")
                        )
                      )
             )
           )
  ),
  
  tabPanel("Genome Browser"),
  
  tabPanel("Gene Checkout", 
           sidebarLayout(
             sidebarPanel(
               tags$textarea(id = "gene_text_area", style = "position: absolute; left: -9999px;"),
               
               actionButton("copy_cart_genes", "Copy Genes to Clipboard"),
               actionButton("clear_cart", "Clear Cart"),
               tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;"),
               tags$h3("Useful Links"),
               div(style = "font-size: 15px;", 
                   tags$ul(
                     tags$li(a(href = "http://bioinformatics.sdstate.edu/go/", 
                               "ShinyGO: Gene Ontology Enrichment Analysis", target = "_blank")),
                     tags$li(a(href = "https://davidbioinformatics.nih.gov/tools.jsp", 
                               "DAVID: Database for Annotation, Visualization and Integrated Discovery", target = "_blank")),
                     tags$li(a(href = "https://pantherdb.org", 
                               "PantherDB: Panther Classification System", target = "_blank"))
                   )
               )
               
             ),
             mainPanel(
               DTOutput("cart_gene_table")
             )
           )
  ),
  
  tabPanel("About",
           fluidPage(
             h2("App Version and Credits"),
             p("Version: 1.0.0"),
             tags$a(href = "https://www.hawkinslab.org", "Hawkins Lab", target = "_blank"),
             p("Developed by: Oliver Brown (ombrown@uw.edu)"),
           )
  )
)

##########

server <- function(input, output, session) {
  
  cart_genes <- reactiveVal(character(0))
  
  if (!exists("comparison_results")) {
    comparison_results <- readRDS("data/comparison_results.rds")
  }
  
  if (!exists("RPKM_data")) {
    RPKM_data <- read.delim("data/logRPKM0625_filter.txt")
  }
  
  if (!exists("phenodata")) {
    phenodata <- read.csv("data/Pheno_data071523.csv")
  }
  
  if (!exists("cnt_data")) {
    cnt_data <- read.delim("data/all_cnt_0625.txt")
  }
  
  observeEvent(input$startAnalysis, {
    runjs('$("#progress-container").show();')
    runjs('$("#progress-bar .progress-bar").css("width", "0%");')
    runjs('$("#progress-text").text("Initializing analysis...");')
    
    withProgress(message = 'Loading...', value = 0, {
      incProgress(0.1, detail = "Updating tissue select inputs...")
      runjs('$("#progress-bar .progress-bar").css("width", "10%");')
      updateSelectInput(session, "tissue_select2", 
                        choices = tissue_names[tissue_names != input$tissue_select1], 
                        selected = isolate(input$tissue_select2))
      
      updateSelectInput(session, "tissue_select1", 
                        choices = tissue_names[tissue_names != input$tissue_select2], 
                        selected = isolate(input$tissue_select1))
      
      incProgress(0.5, detail = "Processing data...")
      runjs('$("#progress-bar .progress-bar").css("width", "30%");')
      filtered_data_combined()
      expressed_genes_by_tissue()
      
      
      incProgress(0.1, detail = "Finalizing...")
      runjs('$("#progress-bar .progress-bar").css("width", "100%");')
      runjs('$("#progress-container").hide();')
    })
  })
  
  # INFO GRAB Tab
  
  output$tissue_display <- renderUI({
    lapply(names(tissue_by_system), function(system) {
      list(
        h3(paste(system, "System")),
        div(style = "display: flex; flex-wrap: wrap; gap: 15px; align-items: center;",
            lapply(tissue_by_system[[system]], function(tissue) {
              div(style = "width: calc(33.33% - 10px); display: flex; align-items: center;",
                  {
                    img_path <- paste0(gsub(" ", "_", tissue), ".png")
                    img_src <- file.path("www", img_path)
                    if (file.exists(img_src)) {
                      img(src = img_path, style = "width: 30px; height: auto; margin-right: 5px;")
                    }
                  },
                  actionLink(inputId = paste0("tissue_", gsub(" ", "_", tissue)), 
                             label = tissue, 
                             style = "font-size: 18px; text-align: left;")
              )
            })
        ),
        tags$hr(style = "height:1px; border:none; color:#300; background-color:#300;")
      )
    })
  })
  
  observe({
    lapply(names(tissue_by_system), function(system) {
      lapply(tissue_by_system[[system]], function(tissue) {
        observeEvent(input[[paste0("tissue_", gsub(" ", "_", tissue))]], {
          samples <- phenodata %>% filter(Tissue == tissue) %>% pull(Sample_name)
          samples_display <- phenodata %>% filter(Tissue == tissue) %>% pull(Sample_name2)
          expressed_genes_count <- expressed_genes_by_tissue() %>%
            filter(Tissue == tissue) %>%
            pull(Expressed_Genes)
          
          tissue_data <- RPKM_data %>% select(samples)
          
          gene_expression_sums <- tissue_data %>% 
            rowSums(na.rm = TRUE) %>%
            as.data.frame() %>%
            setNames("Expression_Sum") %>%
            rownames_to_column(var = "Gene") %>%
            arrange(desc(Expression_Sum))
          
          top_genes <- gene_expression_sums %>% head(5)
          
          showModal(modalDialog(
            title = paste(tissue, "Information"),
            HTML(
              paste(
                "Samples: ", paste(samples_display, collapse = ", "), "<br><hr>",
                "Total Expressed Genes: ", expressed_genes_count, "<br><hr>",
                "Top 5 Most Expressed Genes (RPKM):<br>",
                paste(top_genes$Gene, sep = "", collapse = "<br>")
              )
            ),
            easyClose = TRUE,
            footer = NULL
          ))
        }, ignoreInit = TRUE)
      })
    })
  })
  
  # Data
  
  observe({
    selected_tissue1 <- input$tissue_select1
    updateSelectInput(session, "tissue_select2", 
                      choices = tissue_names[tissue_names != selected_tissue1], 
                      selected = isolate(input$tissue_select2))
  })
  
  observe({
    selected_tissue2 <- input$tissue_select2
    updateSelectInput(session, "tissue_select1", 
                      choices = tissue_names[tissue_names != selected_tissue2], 
                      selected = isolate(input$tissue_select1))
  })
  
  order_val <- reactiveVal(0)
  
  unfiltered_data <- reactive({
    req(input$tissue_select1, input$tissue_select2)
    
    if (paste("Tissue", input$tissue_select1, "vs", input$tissue_select2, sep = "_") %in% names(comparison_results)) {
      comparison_list <- comparison_results[[paste("Tissue", input$tissue_select1, "vs", input$tissue_select2, sep = "_")]]
      order_val(0)
    } 
    else if (paste("Tissue", input$tissue_select2, "vs", input$tissue_select1, sep = "_") %in% names(comparison_results)) {
      comparison_list <- comparison_results[[paste("Tissue", input$tissue_select2, "vs", input$tissue_select1, sep = "_")]]
      order_val(1)
    }
    
    comparison_list <- as.data.frame(comparison_list)
    comparison_list$Gene <- rownames(comparison_list)
    
    
    comparison_list <- comparison_list %>%
      mutate(padj = case_when(
        padj == 0 ~ .Machine$double.xmin,
        TRUE ~ padj
      ))
    
    comparison_list
    
  })
  
  filtered_data_fdr <- reactive({
    data <- unfiltered_data()
    req(data)
    
    if (!is.na(as.numeric(input$selected_fdr))) {
      data <- data %>% filter(padj <= as.numeric(input$selected_fdr))
    }
    
    data
  })
  
  filtered_data_fc <- reactive({
    data <- unfiltered_data()
    req(data)
    
    if (!is.na(as.numeric(input$selected_fc))) {
      data <- data %>% filter(abs(log2FoldChange) >= as.numeric(input$selected_fc))
    }
    
    data
  })
  
  filtered_data_combined <- reactive({
    
    data <- filtered_data_fdr()
    req(data)
    
    
    if (as.numeric(input$selected_fc) != 0) {
      data <- data %>% filter(abs(log2FoldChange) >= as.numeric(input$selected_fc))
    }
    
    data
  })
  
  
  output$title <- renderUI({
    req(input$tissue_select1, input$tissue_select2)
    order <- order_val()
    
    if (order == 0) {
      return(h3(paste(input$tissue_select1, " vs. ", input$tissue_select2, sep = "")))
    }
    
    if (order == 1) {
      return(h3(paste(input$tissue_select2, " vs. ", input$tissue_select1, sep = "")))
    }
    
  })
  
  # Expressed genes counts
  expressed_genes_by_tissue <- reactive({
    # May need to change (or make user able to modify)
    threshold <- 1  
    expressed_genes <- RPKM_data %>%
      rowwise() %>%
      mutate(across(everything(), ~ . > threshold)) %>%
      summarise(across(everything(), sum))
    
    tissue_counts <- phenodata %>%
      filter(Sample_name %in% colnames(RPKM_data)) %>%
      group_by(Tissue) %>%
      summarise(Expressed_Genes = sum(colnames(RPKM_data) %in% Sample_name * expressed_genes))
    
    tissue_counts
  })
  
  output$fold_change_equation <- renderUI({
    req(input$tissue_select1, input$tissue_select2)
    order <- order_val()
    
    if (order == 0) {
      equation <- paste0("<b>log<sub>2</sub> fold change = log<sub>2</sub>(", input$tissue_select1, "/", input$tissue_select2, ")</b>")
    } else {
      equation <- paste0("<b>log<sub>2</sub> fold change = log<sub>2</sub>(", input$tissue_select2, "/", input$tissue_select1, ")</b>")
    }
    
    HTML(equation)
  })
  
  
  
  # Summary of data
  output$summary_info <- renderUI({
    data <- filtered_data_combined()
    order <- order_val()
    req(order)
    data_padj <- filtered_data_fdr()
    data_fc <- filtered_data_fc()
    unfiltered <- unfiltered_data()
    req(data, unfiltered)
    
    total_genes <- nrow(unfiltered)
    total_significant_genes <- nrow(data_padj)
    total_fc_genes <- nrow(data_fc)
    total_combined_genes <- nrow(data)
    
    if (order == 0) {
      upregulated_genes_tissue_1 <- sum(data$log2FoldChange > 0)
      upregulated_genes_tissue_2 <- sum(data$log2FoldChange < 0)
    } else {
      upregulated_genes_tissue_1 <- sum(data$log2FoldChange < 0)
      upregulated_genes_tissue_2 <- sum(data$log2FoldChange > 0)
    }
    
    mean_log2fc <- mean(data$log2FoldChange)
    mean_adj_pvalue <- format(mean(data$padj, na.rm = TRUE), scientific = T, digits = input$num_of_digits)
    
    html_output <- HTML(paste(
      "<table style='width: 50%;'>",
      "<tr><th></th><th></th></tr>",
      "<tr><td>Total genes</td><td>", total_genes, "</td></tr>",
      "<tr><td>Total significant genes</td><td>", total_significant_genes, "</td></tr>",
      "<tr><td>FC > parameter</td><td>", total_fc_genes, "</td></tr>",
      "<tr><td>Significant and FC > parameter</td><td>", total_combined_genes, "</td></tr>",
      "<tr><td colspan='2'><hr></td></tr>",
      "<tr><td>Upregulated genes in ", input$tissue_select1, "</td><td>", upregulated_genes_tissue_1, "</td></tr>",
      "<tr><td>Upregulated genes in ", input$tissue_select2, "</td><td>", upregulated_genes_tissue_2, "</td></tr>",
      "<tr><td colspan='2'><hr></td></tr>",
      "<tr><td>Mean log2 Fold Change</td><td>", round(mean_log2fc, 2), "</td></tr>",
      "<tr><td>Mean Adjusted p-value</td><td>", formatC(mean_adj_pvalue, format = "f", digits = 5), "</td></tr>",
      "<tr><td colspan='2'><hr></td></tr>",
      "</table>"
    ))
    
    return(html_output)
  })
  
  # Search for genes
  
  gene_search_result <- reactiveVal()
  
  observeEvent(input$search_gene, {
    data <- unfiltered_data()
    req(data)
    
    searched_genes <- strsplit(input$searched_gene, ",\\s*")[[1]]
    
    if (length(searched_genes) > 0) {
      gene_data <- data %>% filter(toupper(Gene) %in% toupper(searched_genes))
      not_found_genes <- setdiff(toupper(searched_genes), toupper(gene_data$Gene))
      
      if (length(not_found_genes) > 0) {
        showNotification(paste("The following genes were not found:", paste(not_found_genes, collapse = ", ")), type = "warning")
      }
      
      if (nrow(gene_data) == 0) {
        gene_search_result(data.frame(Statistic = "Genes not found.", Value = ""))
      } else {
        gene_search_result(gene_data)
      }
    }
  })
  
  observeEvent(input$clear_search_main, {
    gene_search_result(NULL)
    updateTextInput(session, "searched_gene", value = "")
  })
  
  observeEvent(input$random_gene, {
    data <- unfiltered_data()
    req(data)
    
    random_genes <- sample(data$Gene, 10)
    updateTextInput(session, "searched_gene", value = paste(random_genes, collapse = ", "))
    gene_data <- data %>% filter(Gene %in% random_genes)
    gene_search_result(gene_data)
  })
  
  output$searched_gene_info <- renderDT({
    gene_data <- gene_search_result()
    
    if (is.null(gene_data)) {
      return(NULL)
    } else if ("Statistic" %in% colnames(gene_data) && gene_data$Statistic == "Genes not found.") {
      return(datatable(gene_data, 
                       options = list(dom = 't', paging = FALSE, ordering = FALSE), rownames = FALSE))
    } else {
      gene_data <- gene_data %>%
        arrange(padj) %>%
        mutate(FalseDiscoveryRate = format(padj, scientific = T, digits = input$num_of_digits), 
               log2FoldChange = format(log2FoldChange, scientific = T, digits = input$num_of_digits)) %>%
        select(Gene, log2FoldChange, FalseDiscoveryRate)
      
      return(datatable(gene_data, 
                       options = list(dom = 't', paging = FALSE, ordering = FALSE), rownames = FALSE))
    }
  })
  
  # Tables
  
  tissue_titles <- reactive({
    if (order_val() == 0) {
      list(
        upregulated = paste("Top 10 Most Upregulated Genes in", input$tissue_select1),
        downregulated = paste("Top 10 Most Upregulated Genes in", input$tissue_select2)
      )
    }
    else {
      list(
        upregulated = paste("Top 10 Most Upregulated Genes in", input$tissue_select2),
        downregulated = paste("Top 10 Most Upregulated Genes in", input$tissue_select1)
      )
    }
  })
  
  output$table_smallest_pvalues <- renderDT({
    data <- filtered_data_combined()
    req(data)
    
    if (nrow(data) == 0) {
      return(datatable(data.frame(Statistic = "No results found for the selected criteria.", Value = "")))
    }
    
    smallest_pvalues <- data %>% 
      arrange(padj) %>%
      mutate(FalseDiscoveryRate = format(padj, scientific = T, digits = input$num_of_digits), 
             log2FoldChange = format(log2FoldChange, scientific = T, digits = input$num_of_digits)) %>%
      head(10) %>%
      select(Gene, log2FoldChange, FalseDiscoveryRate)
    
    datatable(smallest_pvalues, 
              options = list(dom = 't', paging = FALSE, ordering = FALSE), rownames = FALSE)
  })
  
  output$table_largest_fc <- renderDT({
    data <- filtered_data_combined()
    req(data)
    
    if (nrow(data) == 0) {
      return(datatable(data.frame(Statistic = "No results found for the selected criteria.", Value = "")))
    }
    
    largest_fc <- data %>% 
      arrange(desc(abs(log2FoldChange))) %>% 
      mutate(FalseDiscoveryRate = format(padj, scientific = T, digits = input$num_of_digits), 
             log2FoldChange = format(log2FoldChange, scientific = T, digits = input$num_of_digits)) %>%
      head(10) %>%
      select(Gene, log2FoldChange, FalseDiscoveryRate)
    
    datatable(largest_fc, options = list(dom = 't', paging = FALSE, ordering = FALSE), rownames = FALSE)
  })
  
  output$table_upregulated <- renderDT({
    data <- filtered_data_combined()
    req(data)
    
    if (nrow(data) == 0) {
      return(datatable(data.frame(Statistic = "No results found for the selected criteria.", Value = "")))
    }
    
    upregulated <- data %>% 
      arrange(desc(log2FoldChange)) %>% 
      mutate(FalseDiscoveryRate = format(padj, scientific = T, digits = input$num_of_digits), 
             log2FoldChange = format(log2FoldChange, scientific = T, digits = input$num_of_digits)) %>%
      head(10) %>%
      select(Gene, log2FoldChange, FalseDiscoveryRate)
    
    datatable(upregulated, options = list(dom = 't', paging = FALSE, ordering = FALSE), rownames = FALSE)
  })
  
  output$table_downregulated <- renderDT({
    data <- filtered_data_combined()
    req(data)
    
    if (nrow(data) == 0) {
      return(datatable(data.frame(Statistic = "No results found for the selected criteria.", Value = "")))
    }
    
    downregulated <- data %>% 
      arrange(log2FoldChange) %>% 
      mutate(FalseDiscoveryRate = format(padj, scientific = T, digits = input$num_of_digits), 
             log2FoldChange = format(log2FoldChange, scientific = T, digits = input$num_of_digits)) %>%
      head(10) %>%
      select(Gene, log2FoldChange, FalseDiscoveryRate)
    
    datatable(downregulated, options = list(dom = 't', paging = FALSE, ordering = FALSE), rownames = FALSE)
  })
  
  output$table_high_expression <- renderDT({
    data <- filtered_data_combined()
    req(data)
    
    if (nrow(data) == 0) {
      return(datatable(data.frame(Statistic = "No results found for the selected criteria.", Value = "")))
    }
    
    high_expression <- data %>% 
      arrange(desc(baseMean)) %>% 
      head(10) %>%
      select(Gene, baseMean, log2FoldChange, padj)
    
    datatable(high_expression, options = list(dom = 't', paging = FALSE, ordering = FALSE), rownames = FALSE)
  })
  
  output$table_smallest_pvalues <- renderDT({
    data <- filtered_data_combined()
    req(data)
    
    if (nrow(data) == 0) {
      return(datatable(data.frame(Statistic = "No results found for the selected criteria.", Value = "")))
    }
    
    smallest_pvalues <- data %>% 
      arrange(padj) %>% 
      mutate(FalseDiscoveryRate = format(padj, scientific = T, digits = input$num_of_digits), 
             log2FoldChange = format(log2FoldChange, scientific = T, digits = input$num_of_digits)) %>%
      head(10) %>%
      select(Gene, log2FoldChange, FalseDiscoveryRate)
    
    datatable(smallest_pvalues, options = list(dom = 't', paging = FALSE, ordering = FALSE), rownames = FALSE)
  })
  
  output$download_DEA_data <- downloadHandler(
    
    filename = function(){"DEA_data.csv"}, 
    content = function(fname){
      data <- filtered_data_combined()
      data <- data %>%
        select(-Gene)
      write.csv(data, fname)
    }
  )
  
  observeEvent(input$add_all_de_cart, {
    data <- filtered_data_combined()
    req(data)
    
    all_de_genes <- data$Gene
    add_to_cart(all_de_genes)
    
    showNotification(paste(length(all_de_genes), "genes added to the cart."), type = "message")
  })
  
  
  observeEvent(input$add_to_cart_home, {
    genes <- gene_search_result()$Gene
    add_to_cart(genes)
    
    showNotification("genes added to the cart.", type = "message")
  })
  
  observeEvent(input$add_smallest_pvalues_cart, {
    data <- filtered_data_combined()
    req(data)
    
    smallest_pvalues_genes <- data %>% 
      arrange(padj) %>% 
      head(10) %>%
      pull(Gene)
    
    add_to_cart(smallest_pvalues_genes)
    
    showNotification(paste(length(smallest_pvalues_genes), "genes added to the cart from the smallest p-values."), type = "message")
  })
  
  observeEvent(input$add_largest_fc_cart, {
    data <- filtered_data_combined()
    req(data)
    
    largest_fc_genes <- data %>% 
      arrange(desc(abs(log2FoldChange))) %>% 
      head(10) %>%
      pull(Gene)
    
    add_to_cart(largest_fc_genes)
    
    showNotification(paste(length(largest_fc_genes), "genes added to the cart from the largest fold changes."), type = "message")
  })
  
  observeEvent(input$add_upregulated_cart, {
    data <- filtered_data_combined()
    req(data)
    
    upregulated_genes <- data %>% 
      arrange(desc(log2FoldChange)) %>% 
      head(10) %>%
      pull(Gene)
    
    add_to_cart(upregulated_genes)
    
    showNotification(paste(length(upregulated_genes), "genes added to the cart from the upregulated genes."), type = "message")
  })
  
  observeEvent(input$add_downregulated_cart, {
    data <- filtered_data_combined()
    req(data)
    
    downregulated_genes <- data %>% 
      arrange(log2FoldChange) %>% 
      head(10) %>%
      pull(Gene)
    
    add_to_cart(downregulated_genes)
    
    showNotification(paste(length(downregulated_genes), "genes added to the cart from the downregulated genes."), type = "message")
  })
  
  output$collapsePanels <- renderUI({
    bsCollapse(id = "collapseTables", open = "",
               shinyBS::bsCollapsePanel("Ten Smallest p-values", 
                                        DTOutput("table_smallest_pvalues"),
                                        br(),
                                        actionButton("add_smallest_pvalues_cart", "Add to Gene Cart")
               ),
               shinyBS::bsCollapsePanel("Ten Largest Absolute Fold Changes", 
                                        DTOutput("table_largest_fc"),
                                        br(),
                                        actionButton("add_largest_fc_cart", "Add to Gene Cart")
               ),
               shinyBS::bsCollapsePanel(title = tissue_titles()$upregulated, 
                                        DTOutput("table_upregulated"),
                                        br(),
                                        actionButton("add_upregulated_cart", "Add to Gene Cart")
               ),
               shinyBS::bsCollapsePanel(title = tissue_titles()$downregulated, 
                                        DTOutput("table_downregulated"),
                                        br(),
                                        actionButton("add_downregulated_cart", "Add to Gene Cart")
               )
    )
  })
  
  
  ### Volcano Plot
  
  output$title_volcano_plot <- renderUI({
    req(input$tissue_select1, input$tissue_select2)
    h3(paste("Volcano Plot:", input$tissue_select1, "vs.", input$tissue_select2))
  })
  
  search_gene_volcano <- reactiveVal(NULL)
  
  observeEvent(input$search_gene_volcano, {
    search_gene <- input$searched_gene_volcano
    search_gene_volcano(search_gene)
  })
  
  observeEvent(input$clear_search_gene_volcano, {
    search_gene_volcano(NULL)
    updateTextInput(session, "searched_gene_volcano", value = "")
  })
  
  
  generate_volcano_plot <- function(data) {
    selected_fdr <- as.numeric(input$selected_fdr)
    selected_fc <- as.numeric(input$selected_fc)
    order <- order_val()
    searched_gene <- search_gene_volcano()
    
    data <- data %>%
      mutate(Gene = toupper(Gene),
             color = case_when(
               (is.na(selected_fdr) | padj < selected_fdr) & log2FoldChange > selected_fc & order == 0 ~ "Upregulated in Tissue 1",
               (is.na(selected_fdr) | padj < selected_fdr) & log2FoldChange < -selected_fc & order == 0 ~ "Downregulated in Tissue 2",
               (is.na(selected_fdr) | padj < selected_fdr) & log2FoldChange > selected_fc & order == 1 ~ "Upregulated in Tissue 2",
               (is.na(selected_fdr) | padj < selected_fdr) & log2FoldChange < -selected_fc & order == 1 ~ "Downregulated in Tissue 1",
               TRUE ~ "Non-significant"
             ))
    
    if (order == 0) {
      volcano_plot_labels <- c(
        "Upregulated in Tissue 1" = paste("Upregulated in", input$tissue_select1),
        "Downregulated in Tissue 2" = paste("Upregulated in", input$tissue_select2),
        "Non-significant" = "Non-significant"
      )
      label_1 <- input$tissue_select1
      label_2 <- input$tissue_select2
    } else {
      volcano_plot_labels <- c(
        "Upregulated in Tissue 2" = paste("Upregulated in", input$tissue_select2),
        "Downregulated in Tissue 1" = paste("Upregulated in", input$tissue_select1),
        "Non-significant" = "Non-significant"
      )
      label_1 <- input$tissue_select2
      label_2 <- input$tissue_select1
    }
    
    if (input$labeling_option == "basic") {
      top_genes <- data %>%
        filter((is.na(selected_fdr) | padj < selected_fdr) & abs(log2FoldChange) >= selected_fc) %>%
        arrange(padj) %>%
        head(input$n_labels)
    } else {
      top_genes_left <- data %>%
        filter(log2FoldChange < -selected_fc & (is.na(selected_fdr) | padj < selected_fdr)) %>%
        arrange(padj) %>%
        head(input$n_labels_left)
      
      top_genes_right <- data %>%
        filter(log2FoldChange > selected_fc & (is.na(selected_fdr) | padj < selected_fdr)) %>%
        arrange(padj) %>%
        head(input$n_labels_right)
      
      top_genes <- bind_rows(top_genes_left, top_genes_right)
    }
    
    if (input$volcano_background_col == "White") {
      background <- element_blank()
      label_color <- "black"
      searched_gene_color <- "#000000"
      arrow_color <- "black"
      vline_color <- "darkblue"
      hline_color <- "darkred"
    } else { # "Grey"
      background <- element_rect(fill = "grey37",
                                 colour = "grey37",
                                 size = 0.5, linetype = "solid")
      label_color <- "white"
      searched_gene_color <- "#FFFFFF"
      arrow_color <- "white"
      vline_color <- "skyblue"
      hline_color <- "pink"
    }
    
    p <- ggplot(data, aes(x = log2FoldChange, y = -log10(padj))) +
      geom_point(aes(color = color), alpha = 0.5) +
      scale_color_manual(values = c(
        "Upregulated in Tissue 1" = mycolors1[[input$tissue_select1]],
        "Downregulated in Tissue 2" = mycolors1[[input$tissue_select2]],
        "Upregulated in Tissue 2" = mycolors1[[input$tissue_select2]],
        "Downregulated in Tissue 1" = mycolors1[[input$tissue_select1]],
        "Non-significant" = "grey"
      ), labels = volcano_plot_labels) +
      labs(x = "Log2 Fold Change", y = "-log10 Adjusted P-Value", color = "Expression: ") +
      theme_minimal() +
      theme(
        text = element_text(family = "Sen"),
        legend.position = "bottom",
        legend.box = "horizontal",
        legend.box.just = "center",
        legend.text = element_text(size = 14), 
        legend.title = element_text(size = 16),
        legend.background = element_rect(fill = "#f0f0f0", color = "black"),
        legend.key = element_rect(fill = "white", color = "black"),
        legend.key.size = unit(1.5, "lines"), 
        axis.title = element_text(size = 16),
        axis.text = element_text(size = 14),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        panel.background = background,
        axis.line = element_line(color = "black"),
      ) +
      guides(color = guide_legend(override.aes = list(shape = 15, size = 5))) +
      geom_text_repel(data = top_genes, aes(label = Gene), max.overlaps = 1000, color = label_color) +
      geom_hline(yintercept = -log10(selected_fdr), linetype = "dashed", color = hline_color, alpha = 1/3) +
      geom_vline(xintercept = c(-selected_fc, selected_fc), linetype = "dashed", color = vline_color, alpha = 1/3) +
      annotate("text", x = 0 + max(data$log2FoldChange) * 0.5, 
               y = max(-log10(data$padj), na.rm = TRUE) * 1.2, 
               label = label_1, size = 5, hjust = 0.5, color = label_color) +
      annotate("text", x = 0 - max(data$log2FoldChange) * 0.5, 
               y = max(-log10(data$padj), na.rm = TRUE) * 1.2, 
               label = label_2, size = 5, hjust = 0.5, color = label_color) +
      annotate("segment", x = 0 + max(data$log2FoldChange) * 0.4, 
               y = max(-log10(data$padj), na.rm = TRUE) * 1.175, 
               xend = 0 + max(data$log2FoldChange) * 0.6, 
               yend = max(-log10(data$padj), na.rm = TRUE) * 1.175,
               arrow = arrow(length = unit(0.3, "cm")), color = arrow_color) +
      annotate("segment", x = 0 - max(data$log2FoldChange) * 0.4, 
               y = max(-log10(data$padj), na.rm = TRUE) * 1.175, 
               xend = 0 - max(data$log2FoldChange) * 0.6, 
               yend = max(-log10(data$padj), na.rm = TRUE) * 1.175,
               arrow = arrow(length = unit(0.3, "cm")), color = arrow_color)
    
    if (!is.null(searched_gene) && toupper(searched_gene) %in% data$Gene) {
      p <- p + geom_point(data = data %>% filter(Gene == toupper(searched_gene)), 
                          aes(x = log2FoldChange, y = -log10(padj)), color = searched_gene_color, size = 3)
    } else if (!is.null(searched_gene) && !(toupper(searched_gene) %in% data$Gene)) {
      showNotification(paste("Gene", searched_gene, "not found."), type = "warning")
    }
    
    return(p)
  }
  
  
  output$volcano_plot <- renderPlot({
    data <- unfiltered_data()
    req(data)
    
    p <- generate_volcano_plot(data)
    print(p)
  })
  
  output$download_volcano_plot <- downloadHandler(
    filename = function() { paste('volcano.png') },
    content = function(file) {
      data <- unfiltered_data()
      req(data)
      
      p <- generate_volcano_plot(data)
      ggsave(file, plot = p, width = 20, height = 13, units = "in", dpi = 300, bg = "white")
    }
  )
  
  observeEvent(input$add_to_cart_volcano, {
    data <- unfiltered_data()
    gene <- search_gene_volcano()
    
    if (!is.null(gene) && gene != "") {
      # Convert gene to uppercase for comparison
      gene_upper <- toupper(gene)
      
      # Check if the gene exists in the dataset
      if (gene_upper %in% toupper(data$Gene)) {
        add_to_cart(gene_upper)
        showNotification("Gene added to the cart.", type = "message")
      } else {
        showNotification(paste("Gene", gene, "not found."), type = "warning")
      }
    } else {
      showNotification("No gene specified.", type = "warning")
    }
  })
  
  ### Heatmap
  
  rpkm_filtered <- reactiveVal()
  ordered_genes <- reactiveVal(NULL) 
  
  output$title_heatmap <- renderUI({
    req(input$tissue_select1, input$tissue_select2)
    h3(paste("Heatmap:", input$tissue_select1, "vs.", input$tissue_select2))
  })
  
  search_genes_heatmap <- reactiveVal(NULL)
  
  observeEvent(input$search_gene_heatmap, {
    data <- unfiltered_data()
    req(data)
    
    searched_genes <- strsplit(input$search_genes_heatmap, ",\\s*")[[1]]
    
    if (length(searched_genes) > 0) {
      gene_data <- data %>% filter(toupper(Gene) %in% toupper(searched_genes))
      not_found_genes <- setdiff(toupper(searched_genes), toupper(gene_data$Gene))
      
      if (length(not_found_genes) > 0) {
        showNotification(paste("The following genes were not found:", paste(not_found_genes, collapse = ", ")), type = "warning")
      }
      
      if (nrow(gene_data) > 0) {
        search_genes_heatmap(gene_data$Gene)
      }
    }
  })
  
  observeEvent(input$random_gene_heatmap, {
    data <- filtered_data_combined()
    req(data)
    
    random_genes <- sample(data$Gene, 10)
    search_genes_heatmap(random_genes)
    updateTextInput(session, "search_genes_heatmap", value = paste(random_genes, collapse = ", "))
  })
  
  observeEvent(input$clear_search, {
    search_genes_heatmap(NULL)
    updateTextInput(session, "search_genes_heatmap", value = "")
  })
  
  generate_plot <- function(data, search_genes) {
    if (input$sort_by == "FDR") {
      data <- data %>% arrange(padj)
    } else {
      data <- data %>% arrange(desc(abs(log2FoldChange)))
    }
    
    search_genes <- search_genes_heatmap()
    
    if (!is.null(search_genes) && length(search_genes) > 0) {
      top_genes <- search_genes
    } else {
      top_genes <- data %>% 
        head(as.numeric(input$top_n)) %>% 
        pull(Gene)
    }
    
    selected_tissues <- c(input$tissue_select1, input$tissue_select2)
    
    samples <- phenodata %>%
      filter(Tissue %in% selected_tissues) %>%
      pull(Sample_name)
    
    rpkm_filtered <- RPKM_data %>%
      filter(rownames(.) %in% top_genes)
    
    samples <- samples[samples %in% colnames(rpkm_filtered)]
    
    rpkm_filtered <- rpkm_filtered[, samples, drop = FALSE]
    
    colnames(rpkm_filtered) <- gsub("Sample_", "", colnames(rpkm_filtered))
    
    annotation <- phenodata %>%
      filter(Sample_name %in% samples) %>%
      select(Sample_name, Tissue) %>%
      mutate(Sample_name = gsub("Sample_", "", Sample_name)) %>%
      column_to_rownames(var = "Sample_name")
    
    present_tissues <- unique(annotation$Tissue)
    annotation_colors <- list(Tissue = mycolors1[present_tissues])
    
    n_genes <- nrow(rpkm_filtered)
    fontsize_row <- ifelse(n_genes > 50, 7, 14)
    show_rownames <- n_genes <= 100
    
    color_scale <- switch(input$color_scale,
                          "blue_white_red" = c("royalblue", "white", "firebrick3"),
                          "green_black_red" = c("springgreen2", "black", "firebrick2"),
                          "purple_white_green" = c("purple", "white", "springgreen4"),
                          "viridis" = viridis(100),
                          "plasma" = plasma(100),
                          "cividis" = cividis(100),
                          "inferno" = inferno(100))
    
    rpkm_filtered(rpkm_filtered)
    
    heatmap <- pheatmap(
      rpkm_filtered, 
      scale = "row",
      cluster_rows = (n_genes > 1), 
      cluster_cols = TRUE, 
      show_rownames = show_rownames, 
      show_colnames = TRUE,
      annotation_col = annotation,
      annotation_colors = annotation_colors,
      color = colorRampPalette(color_scale)(100),
      legend_labels = c("low", "medium", "high"),
      cutree_cols = 2,
      cutree_rows = if (n_genes > 1) 2 else 1,
      angle_col = 45,
      fontsize = 13,
      fontsize_col = 14,
      fontsize_row = fontsize_row,
      width = 30,
      treeheight_row = 0,
      treeheight_col = 0
    )
    
    ordered_genes(heatmap$tree_row$labels[heatmap$tree_row$order])
    
    rpkm_filtered(rpkm_filtered)
    
    heatmap
  }
  
  output$heatmap_plot <- renderPlot({
    data <- filtered_data_combined()
    req(data)
    search_genes <- search_genes_heatmap()
    generate_plot(data, search_genes)
  })
  
  observeEvent(input$heatmap_brush, {
    brush_data <- input$heatmap_brush
    
    if (!is.null(brush_data)) {
      # Get the ordered genes from the heatmap
      heatmap_rows <- ordered_genes()
      total_height <- length(heatmap_rows)
      
      # Calculate the indices based on brush coordinates, reversing the order
      start_row_index <- total_height - ceiling(brush_data$ymin / (1 / total_height)) + 1
      end_row_index <- total_height - ceiling(brush_data$ymax / (1 / total_height)) + 1
      
      # Ensure indices are within the valid range
      start_row_index <- max(min(start_row_index, total_height), 1)
      end_row_index <- max(min(end_row_index, total_height), 1)
      
      # Select the correct genes based on reversed indices
      selected_rows <- heatmap_rows[end_row_index:start_row_index]
      
      # Store the selected genes in reactiveVal
      selected_genes(selected_rows)
      
      # Show modal with the selected genes and add to cart button
      showModal(modalDialog(
        title = "Selected Genes",
        paste("You have selected:", paste(selected_rows, collapse = ", ")),
        footer = tagList(
          actionButton("add_to_cart_tissue_specific_modal", "Add to Gene Cart"),
          modalButton("Close")
        ),
        easyClose = TRUE
      ))
    }
  })
  
  
  
  output$download_heatmap_plot <- downloadHandler(
    filename = function() {paste('heatmap.png')},
    content = function(file) {
      data <- filtered_data_combined()
      req(data)
      
      png(file, width = 18, height = 13, units = "in", res = 300)
      generate_plot(data, search_genes_heatmap())
      dev.off()
    }
  )
  
  # Add to Cart functionality for the Heatmap tab
  observeEvent(input$add_to_cart_heatmap, {
    data <- filtered_data_combined()
    req(data)
    
    # Get the genes currently displayed in the heatmap
    search_genes <- search_genes_heatmap()
    
    if (!is.null(search_genes) && length(search_genes) > 0) {
      top_genes <- search_genes
    } else {
      top_genes <- data %>% 
        head(as.numeric(input$top_n)) %>% 
        pull(Gene)
    }
    
    add_to_cart(top_genes)
    
    showNotification("genes added to the cart.", type = "message")
  })
  
  
  ### Correlation Plot
  
  current_correlation_plot <- reactiveVal(NULL)
  
  output$title_cor_plot <- renderUI({
    req(input$tissue_select1, input$tissue_select2)
    
    if (input$correlation_mode == "selected") {
      h3(paste("Correlation Plot:", input$tissue_select1, "and", input$tissue_select2))
    } else if (input$correlation_mode == "system") {
      h3(paste("Correlation Plot:", input$correlation_system, "system"))
    }
  })
  
  observe({
    
    selected_samples <- NULL
    
    if (input$correlation_mode == "selected") {
      selected_tissues <- c(input$tissue_select1, input$tissue_select2)
      selected_samples <- phenodata %>%
        filter(Tissue %in% selected_tissues) %>%
        arrange(Tissue) %>%
        pull(Sample_name2)
    } else if (input$correlation_mode == "system") {
      selected_system <- input$correlation_system
      selected_tissues <- phenodata %>%
        filter(System == selected_system) %>%
        pull(Tissue) %>%
        unique()
      
      selected_samples <- phenodata %>%
        filter(Tissue %in% selected_tissues) %>%
        arrange(Tissue) %>%
        pull(Sample_name2)
    }
    
    req(selected_samples)
    
    N <- 1000
    gene_variability <- apply(RPKM_data, 1, var)
    top_genes <- names(sort(gene_variability, decreasing = TRUE)[1:N])
    
    sample_names1 <- phenodata$Sample_name[match(selected_samples, phenodata$Sample_name2)]
    
    if (input$use_filtered_data) {
      filtered_genes <- rownames(filtered_data_combined())
      filtered_RPKM <- RPKM_data[rownames(RPKM_data) %in% filtered_genes, , drop = FALSE]
    } else {
      filtered_RPKM <- RPKM_data
    }
    
    filtered_RPKM <- filtered_RPKM[rownames(filtered_RPKM) %in% top_genes, , drop = FALSE]
    
    selected_data <- filtered_RPKM[, sample_names1, drop = FALSE]
    
    colnames(selected_data) <- selected_samples
    
    sample_tissues <- phenodata$Tissue[match(selected_samples, phenodata$Sample_name2)]
    
    create_diag <- function(data, mapping, ...) {
      varname <- as_label(mapping$x)
      tissue <- phenodata$Tissue[match(varname, phenodata$Sample_name2)]
      color <- mycolors1[tissue]
      ggplot(data = data, mapping = mapping) +
        geom_density() +
        theme_void() +
        theme(
          panel.background = element_rect(fill = color, colour = "black", size = 2)
        )
    }
    
    custom_ggally_cor <- function(data, mapping, ...) {
      x <- eval_data_col(data, mapping$x)
      y <- eval_data_col(data, mapping$y)
      cor_value <- cor(x, y, use = "complete.obs")
      label <- format(cor_value, digits = 2)
      
      tile_data <- data.frame(cor_value = cor_value)
      
      ggplot(data = tile_data) +
        geom_tile(aes(x = 0.5, y = 0.5, fill = cor_value), width = -Inf, height = Inf, alpha = 1) +
        geom_text(aes(x = 0.5, y = 0.5, label = label), size = 5.5, hjust = 0.5, vjust = 0.5) +
        scale_fill_gradient2(low = "blue", mid = "white", high = "firebrick3", limits = c(-1, 1), na.value = NA) +
        theme_void() +
        theme(legend.position = "none")
    }
    
    create_upper <- function(data, mapping, ...) {
      custom_ggally_cor(data = data, mapping = mapping) +
        theme_minimal() +
        theme(
          plot.title = element_blank(),
          axis.title = element_blank()
        )
    }
    
    # Store the plot object in a reactive value
    correlation_plot <- reactive({
      ggpairs(
        data = as.data.frame(selected_data),
        upper = list(continuous = create_upper),
        diag = list(continuous = create_diag),
        lower = list(continuous = wrap("points", color = "black", alpha = 0.5))
      ) +
        theme(
          legend.position = "none",
          legend.box = "horizontal",
          legend.box.just = "center",
          legend.text = element_text(size = 14), 
          legend.title = element_text(size = 16),
          legend.background = element_rect(fill = "#f0f0f0", color = "black"),
          legend.key = element_rect(fill = "white", color = "black"),
          legend.key.size = unit(1.5, "lines")
        )
    })
    
    if (ncol(selected_data) > 1) {
      output$correlation_plot <- renderPlot({
        p <- correlation_plot()
        
        present_tissues <- unique(sample_tissues)
        num_tissues <- length(present_tissues)
        
        # Create a data frame for legend positioning
        legend_data <- data.frame(
          Tissue = present_tissues,
          Color = mycolors1[present_tissues],
          n = num_tissues:1
        )
        
        legend_data$Tissue <- trimws(legend_data$Tissue)
        
        # Calculate the length of each label
        legend_data$label_length <- nchar(legend_data$Tissue)
        
        # Calculate the maximum label length
        max_label_length <- max(nchar(legend_data$Tissue))
        
        legend_data$x_pos <- 1.90 - (legend_data$label_length / max_label_length) * 0*3  # Adjust 0.3 as needed
        
        legend <- ggplot(legend_data, aes(x = 0.9, y = n)) +
          geom_point(aes(color = Tissue), size = 5, shape = 15) +
          geom_text(aes(label = Tissue, x = x_pos), 
                    vjust = 0.5, size = 5, color = "black") +
          scale_color_manual(values = mycolors1[present_tissues]) +
          xlim(0.5, 3) +  # Adjust x limits for spacing
          ylim(-2*length(present_tissues), length(present_tissues)) +  # Adjust y limits for vertical spacing
          theme_void() +
          theme(
            legend.position = "none",
            plot.margin = margin(0, 0, 0, 0, "pt")
          )
        
        
        
        
        grid.newpage()
        pushViewport(viewport(layout = grid.layout(1, 2, widths = unit(c(4, 1), "null"))))
        print(p, vp = viewport(layout.pos.row = 1, layout.pos.col = 1))
        print(legend, vp = viewport(layout.pos.row = 1, layout.pos.col = 2))
        
        library(grid)
        library(gridExtra)
        
        # Create the combined plot object (without embedding the legend in the plot)
        combined_plot <- arrangeGrob(p, legend, ncol = 1, heights = c(4, 1))
        
        # Save the combined plot to the reactive value
        current_correlation_plot(combined_plot)
        
        
        
      })
      
      output$download_correlation_plot <- downloadHandler(
        filename = function() {
          paste("Correlation", ".png", sep = "")
        },
        content = function(file) {
          ggsave(file, plot = correlation_plot(), device = "png", width = 14, height = 10)
        }
      )
    }
  })
  
  
  
  
  
  
  
  ### PCA
  
  output$title_pca_plot <- renderUI({
    req(input$tissue_select1, input$tissue_select2)
    
    if (input$pca_mode == "selected") {
      h3(paste("PCA:", input$tissue_select1, "and", input$tissue_select2))
    } else if (input$pca_mode == "system") {
      h3(paste("PCA:", input$pca_system_select, "System"))
    }
    else if (input$pca_mode == "all") {
      h3(paste("PCA: All Tissues"))
    }
  })
  
  observe({
    selected_samples <- NULL
    
    if (input$pca_mode == "selected") {
      selected_tissues <- c(input$tissue_select1, input$tissue_select2)
      selected_samples <- phenodata %>%
        filter(Tissue %in% selected_tissues) %>%
        arrange(Tissue) %>%
        pull(Sample_name2)
      
    } else if (input$pca_mode == "system") {
      selected_system <- input$pca_system_select
      selected_tissues <- phenodata %>%
        filter(System == selected_system) %>%
        pull(Tissue) %>%
        unique()
      
      selected_samples <- phenodata %>%
        filter(Tissue %in% selected_tissues) %>%
        arrange(Tissue) %>%
        pull(Sample_name2)
      
    } else if (input$pca_mode == "all") {
      selected_samples <- phenodata %>%
        pull(Sample_name2)
    }
    
    req(selected_samples)
    
    N <- 1000
    gene_variability <- apply(RPKM_data, 1, var)
    top_genes <- names(sort(gene_variability, decreasing = TRUE)[1:N])
    
    sample_names1 <- phenodata$Sample_name[match(selected_samples, phenodata$Sample_name2)]
    
    filtered_RPKM <- RPKM_data[rownames(RPKM_data) %in% top_genes, , drop = FALSE]
    
    selected_data <- filtered_RPKM[, sample_names1, drop = FALSE]
    
    colnames(selected_data) <- selected_samples
    
    sample_tissues <- phenodata$Tissue[match(selected_samples, phenodata$Sample_name2)]
    
    output$pca_plot <- renderPlot({
      
      pca_data <- prcomp(t(selected_data), scale. = F)
      pca_df <- as.data.frame(pca_data$x)
      pca_df$Sample <- rownames(pca_df)
      pca_df <- merge(pca_df, phenodata, by.x = "Sample", by.y = "Sample_name2")
      
      ggplot_pca <- ggplot(pca_df, aes(x = PC1, y = PC2, color = Tissue, shape = System)) +
        geom_point(size = 4) +
        geom_polygon(data = pca_df, aes(x = PC1, y = PC2, group = Tissue, fill = Tissue), alpha = 0.2) +
        scale_color_manual(values = mycolors1) +
        labs(x = "Principal Component 1", y = "Principal Component 2") +
        theme_minimal()
      
      print(ggplot_pca)
      
      output$download_pca_plot <- downloadHandler(
        filename = function() {
          paste("PCA", ".png", sep = "")
        },
        content = function(file) {
          ggsave(file, plot = ggplot_pca, device = "png", width = 10, height = 8, bg = "white")
        }
      )
    })
  })
  
  
  
  
  ### barplots
  
  barplot_data <- reactiveVal(NULL)
  current_plot <- reactiveVal(NULL)
  
  # Add a reactive value to keep track of the current title
  current_title <- reactiveVal("")
  
  common_genes <- function(tissue1, tissue2, RPKM_data, phenodata) {
    # Filter phenotype data to get the samples belonging to the selected tissues
    tissue1_samples <- phenodata %>%
      filter(Tissue == tissue1) %>%
      pull(Sample_name)
    
    tissue2_samples <- phenodata %>%
      filter(Tissue == tissue2) %>%
      pull(Sample_name)
    
    # Select the columns in RPKM_data that match the filtered sample names
    selected_samples <- RPKM_data %>%
      select(all_of(c(tissue1_samples, tissue2_samples)))
    
    # Calculate the average expression across the selected tissues for each gene
    top_genes <- selected_samples %>%
      rowMeans(na.rm = TRUE) %>%
      sort(decreasing = TRUE) %>%
      head(5) %>%
      names()
    
    return(top_genes)
  }
  
  observeEvent(input$differential_analysis_tab, {
    if (input$differential_analysis_tab == "Gene Expression Chart") {

      req(input$tissue_select1, input$tissue_select2)
      
      top_genes <- common_genes(input$tissue_select1, input$tissue_select2, RPKM_data, phenodata)
      
      updateTextInput(session, "barplot_searched_gene", value = paste(top_genes, collapse = ", "))
      updateBarplotData(top_genes)

      current_title(paste("Showing Top 5 Commonly Expressed Genes in", input$tissue_select1, "and", input$tissue_select2))
    }
  })
  
  
  observeEvent(input$search_gene_barplot, {
    req(input$barplot_searched_gene)
    
    searched_genes <- strsplit(input$barplot_searched_gene, ",\\s*")[[1]]
    searched_genes <- toupper(searched_genes)
    
    found_genes <- intersect(toupper(rownames(RPKM_data)), searched_genes)
    
    if (length(found_genes) == 0) {
      showNotification("No genes found.", type = "warning")
      barplot_data(NULL)
      current_plot(NULL)
      return(NULL)
    }
    
    updateBarplotData(found_genes)
    
    # Update the title
    current_title("Showing Searched Genes")
  })
  
  observeEvent(input$random_gene_barplot, {
    random_genes <- sample(rownames(RPKM_data), 5)
    updateTextInput(session, "barplot_searched_gene", value = paste(random_genes, collapse = ", "))
    updateBarplotData(random_genes)
    
    # Update the title
    current_title("Showing searched genes")
  })
  
  observeEvent(input$clear_search_gene_barplot, {
    updateTextInput(session, "barplot_searched_gene", value = "")
    barplot_data(NULL)
    current_plot(NULL)
    output$dynamic_barplot_output <- renderUI({
      NULL
    })
    
    current_title(paste("Showing Top 5 Commonly Expressed Genes in", input$tissue_select1, "and", input$tissue_select2))
    
    req(input$tissue_select1, input$tissue_select2)
    
    top_genes <- common_genes(input$tissue_select1, input$tissue_select2, RPKM_data, phenodata)
    
    updateTextInput(session, "barplot_searched_gene", value = paste(top_genes, collapse = ", "))
    updateBarplotData(top_genes)
  })
  
  # Render the title in the UI
  output$dynamic_title <- renderUI({
    h2(current_title(), style = "text-align: center; margin-top: 20px;")
  })
  
  updateBarplotData <- function(genes) {
    if (is.null(genes) || length(genes) == 0) {
      showNotification("No valid genes to plot.", type = "error")
      output$dynamic_barplot_output <- renderUI({
        h4("No data available for the selected genes and tissues.", style = "text-align: center; color: red;")
      })
      return(NULL)
    }
    
    selected_data <- RPKM_data[genes, , drop = FALSE]
    
    if (nrow(selected_data) == 0 || ncol(selected_data) == 0) {
      showNotification("No valid data available for the selected genes and tissues.", type = "error")
      output$dynamic_barplot_output <- renderUI({
        h4("No data available for the selected genes and tissues.", style = "text-align: center; color: red;")
      })
      return(NULL)
    }
    
    selected_data$Gene <- rownames(selected_data)
    
    melted_data <- reshape2::melt(selected_data)
    tissue_info <- phenodata[match(melted_data$variable, phenodata$Sample_name), "Tissue"]
    melted_data$Tissue <- tissue_info
    
    # Ensure that melted_data is valid for plotting
    if (nrow(melted_data) == 0 || all(is.na(melted_data$Tissue))) {
      showNotification("No valid data available for the selected genes and tissues.", type = "error")
      output$dynamic_barplot_output <- renderUI({
        h4("No data available for the selected genes and tissues.", style = "text-align: center; color: red;")
      })
      return(NULL)
    }
    
    average_expression_per_tissue <- melted_data %>%
      group_by(Gene, Tissue) %>%
      summarize(average_value = mean(value, na.rm = TRUE), .groups = 'drop') %>%
      ungroup()
    
    # Ensure the data used for plotting is valid
    if (nrow(average_expression_per_tissue) == 0) {
      showNotification("No valid data available for the selected genes and tissues.", type = "error")
      output$dynamic_barplot_output <- renderUI({
        h4("No data available for the selected genes and tissues.", style = "text-align: center; color: red;")
      })
      return(NULL)
    }
    
    barplot_data(average_expression_per_tissue)
    
    output$dynamic_barplot_output <- renderUI({
      plot_height <- 300 + 300 * length(genes)
      plotOutput("barplot_plot", height = paste0(plot_height, "px"))
    })
    
    output$barplot_plot <- renderPlot({
      p <- ggplot(average_expression_per_tissue, aes(x = Tissue, y = average_value, fill = Tissue)) +
        geom_bar(stat = "identity", color = "black", width = 0.8) + 
        facet_wrap(~ Gene, scales = "free", ncol = 1) +
        theme_minimal() +
        theme(
          axis.text.x = element_text(angle = 45, hjust = 1, size = 12),
          panel.spacing = unit(0.5, "lines"), 
          panel.border = element_rect(color = "black", fill = NA, size = 0.5),
          strip.background = element_blank(),
          strip.placement = "outside",
          panel.background = element_blank(),
          strip.text = element_text(size = 14, face = "bold"),
          legend.position = "none"
        ) +
        scale_fill_manual(values = mycolors1) +
        labs(y = "Average Expression (Avg. RPKM)", x = "Tissue")
      
      current_plot(p)
      print(p)
    })
  }
  
  
  output$download_barplot <- downloadHandler(
    filename = function() {
      paste("Barplot", ".png", sep = "")
    },
    content = function(file) {
      ggsave(file, plot = current_plot(), device = "png", width = 10, height = length(barplot_data()) * 4, bg = "white")
    }
  )
  
  observeEvent(input$add_to_cart_barplot, {
    genes <- unique(barplot_data()$Gene)
    add_to_cart(genes)
    
    showNotification("genes added to the cart.", type = "message")
  })
  
  
  ##########################
  
  # Tissue specific analysis
  
  selected_genes <- reactiveVal()
  rpkm_heatmap_ordered <- reactiveVal()
  ordered_tissue_specific_genes <- reactiveVal(NULL)
  
  
  observeEvent(input$system_choice, {
    all_systems_selected <- setdiff(unique(phenodata$System), "All Systems")
    
    if ("All Systems" %in% input$system_choice && length(input$system_choice) > 1) {
      updateCheckboxGroupInput(session, "system_choice", 
                               selected = setdiff(input$system_choice, "All Systems"))
    }
    
    if (all(all_systems_selected %in% input$system_choice)) {
      updateCheckboxGroupInput(session, "system_choice", 
                               selected = "All Systems")
    }
  })
  
  output$tissue_selection_ui <- renderUI({
    req(input$system_choice)
    
    if ("All Systems" %in% input$system_choice) {
      tissues <- unique(phenodata$Tissue)
    } else {
      tissues <- unique(phenodata$Tissue[phenodata$System %in% input$system_choice])
    }
    
    checkboxGroupInput("tissue_choice", "Select Tissue(s):", choices = tissues, selected = tissues)
  })
  
  output$tissue_specific_heatmap <- renderPlot({
    req(input$tissue_choice)
    current_plot(NULL)
    
    selected_tissues <- input$tissue_choice
    
    samples <- phenodata %>%
      filter(Tissue %in% selected_tissues) %>%
      pull(Sample_name)
    
    samples <- samples[samples %in% colnames(RPKM_data)]
    
    rpkm_filtered <- RPKM_data[, samples, drop = FALSE]
    
    non_zero_genes <- apply(rpkm_filtered, 1, function(row) !all(row == 0))
    rpkm_filtered <- rpkm_filtered[non_zero_genes, ]
    
    colnames(rpkm_filtered) <- gsub("Sample_", "", colnames(rpkm_filtered))
    
    annotation <- phenodata %>%
      filter(Sample_name %in% samples) %>%
      select(Sample_name, Tissue) %>%
      mutate(Sample_name = gsub("Sample_", "", Sample_name)) %>%
      column_to_rownames(var = "Sample_name")
    
    tau_scores <- calcTau(rpkm_filtered)
    
    filtered_tau <- tau_scores[tau_scores$tau >= 0.8, ]
    rpkm_filtered <- rpkm_filtered[rownames(rpkm_filtered) %in% rownames(filtered_tau), ]
    
    rpkm_tau <- as.matrix(rpkm_filtered)
    
    variances <- apply(t(rpkm_tau), 2, var)
    
    top_genes <- order(variances, decreasing = TRUE)[1:input$n_variable_genes]
    rpkm_heatmap <- rpkm_tau[top_genes,]
    
    annotation <- phenodata %>%
      filter(Sample_name %in% samples) %>%
      select(Sample_name, Tissue, System) %>%
      mutate(Sample_name = gsub("Sample_", "", Sample_name)) %>%
      column_to_rownames(var = "Sample_name")
    
    present_tissues <- unique(annotation$Tissue)
    
    annotation_colors <- list(
      Tissue = mycolors1[unique(annotation$Tissue)],
      System = mycolors2[unique(annotation$System)]
    )
    
    n_genes <- nrow(rpkm_tau)
    
    color_scale_tissue_specific <- switch(input$color_scale_tissue_specific,
                                          "blue_white_red" = c("royalblue", "white", "firebrick3"),
                                          "green_black_red" = c("springgreen2", "black", "firebrick2"),
                                          "purple_white_green" = c("purple", "white", "springgreen4"),
                                          "cyan_yellow_purple" = c("purple", "lightyellow", "blue"),
                                          "viridis" = viridis(100),
                                          "plasma" = plasma(100),
                                          "cividis" = cividis(100),
                                          "inferno" = inferno(100))
    
    annotation_order <- annotation %>%
      arrange(System, Tissue) %>%
      rownames()
    
    rpkm_heatmap_ordered <- rpkm_heatmap[, annotation_order]
    
    rpkm_heatmap_ordered(rpkm_heatmap_ordered)
    
    
    heatmap <- pheatmap(
      rpkm_heatmap_ordered, 
      annotation_col = annotation[annotation_order, ],
      annotation_colors = annotation_colors,
      color = colorRampPalette(color_scale_tissue_specific)(100),
      clustering_distance_rows = "correlation",
      clustering_distance_cols = "correlation",
      clustering_method = "complete",
      scale = "row",
      drop_levels = TRUE,
      border_color = "NA",
      cluster_rows = TRUE,
      cluster_cols = FALSE,
      show_rownames = FALSE,  # Ensure row names are shown for interaction
      show_colnames = TRUE,
      legend_breaks = c(-6, 0, 6),
      legend_labels = c("low", "medium", "high"),
      legend = TRUE,
      main = "Tissue-Specific (TS) Transcript Isoforms",
      angle_col = 45,
      fontsize = 12,
      fontsize_col = 9,
      treeheight_col = 0
    )
    
    ordered_tissue_specific_genes(heatmap$tree_row$labels[heatmap$tree_row$order])
    
    current_plot(heatmap)
    
    print(heatmap)
  })
  
  
  observeEvent(input$heatmap_tissue_specific_brush, {
    brush_data <- input$heatmap_tissue_specific_brush
    
    if (!is.null(brush_data)) {
      # Get the ordered genes from the heatmap
      heatmap_rows <- ordered_tissue_specific_genes()
      total_height <- length(heatmap_rows)
      
      # Calculate the indices based on brush coordinates, reversing the order
      start_row_index <- total_height - ceiling(brush_data$ymin / (1 / total_height)) + 1
      end_row_index <- total_height - ceiling(brush_data$ymax / (1 / total_height)) + 1
      
      # Ensure indices are within the valid range
      start_row_index <- max(min(start_row_index, total_height), 1)
      end_row_index <- max(min(end_row_index, total_height), 1)
      
      # Select the correct genes based on reversed indices
      selected_rows <- heatmap_rows[end_row_index:start_row_index]
      
      # Store the selected genes in reactiveVal
      selected_genes(selected_rows)
      
      # Show modal with the selected genes and add to cart button
      showModal(modalDialog(
        title = "Selected Genes",
        paste("You have selected:", paste(selected_rows, collapse = ", ")),
        footer = tagList(
          actionButton("add_to_cart_tissue_specific_modal", "Add to Gene Cart"),
          modalButton("Close")
        ),
        easyClose = TRUE
      ))
    }
  })
  
  output$download_tissue_specific_heatmap <- downloadHandler(
    filename = function() {
      paste("Tissue_specific_heatmap", ".png", sep = "")
    },
    content = function(file) {
      ggsave(file, plot = current_plot(), device = "png", width = 10, height = 11, bg = "white")
    }
  )
  
  observeEvent(input$add_to_cart_tissue_specific_modal, {
    genes_to_add <- selected_genes()
    
    if (!is.null(genes_to_add)) {
      add_to_cart(genes_to_add)
      
      showNotification("Genes added to the cart.", type = "message")
      
      removeModal()
    }
  })
  
  
  observeEvent(input$add_to_cart_tissue_specific_heatmap, {
    req(input$tissue_choice, input$n_variable_genes)
    
    selected_tissues <- input$tissue_choice
    
    samples <- phenodata %>%
      filter(Tissue %in% selected_tissues) %>%
      pull(Sample_name)
    
    samples <- samples[samples %in% colnames(RPKM_data)]
    
    rpkm_filtered <- RPKM_data[, samples, drop = FALSE]
    
    non_zero_genes <- apply(rpkm_filtered, 1, function(row) !all(row == 0))
    rpkm_filtered <- rpkm_filtered[non_zero_genes, ]
    
    variances <- apply(t(rpkm_filtered), 2, var)
    top_genes <- order(variances, decreasing = TRUE)[1:input$n_variable_genes]
    top_genes <- rownames(rpkm_filtered)[top_genes]
    
    add_to_cart(top_genes)
    
    showNotification("Genes added to the cart.", type = "message")
  })
  
  
  ##############
  
  # Genome Browser
  
  
  
  
  ##############
  
  # Gene cart
  
  add_to_cart <- function(new_genes) {
    current_cart <- cart_genes()
    updated_cart <- unique(c(current_cart, new_genes))
    cart_genes(updated_cart)
  }
  
  observeEvent(input$clear_cart, {
    showModal(
      modalDialog(
        title = "Confirm Clear Cart",
        "Are you sure you want to clear all genes from the cart?",
        easyClose = FALSE,
        footer = tagList(
          modalButton("Cancel"),
          actionButton("confirm_clear_cart", "Yes, Clear Cart")
        )
      )
    )
  })
  
  observeEvent(input$confirm_clear_cart, {
    cart_genes(character(0))
    removeModal()
    showNotification("Gene cart has been cleared.", type = "message")
  })
  
  output$cart_gene_table <- renderDT({
    genes <- cart_genes()
    if (length(genes) == 0) {
      return(datatable(data.frame(Gene = "Cart is empty")))
    } else {
      return(datatable(data.frame(Gene = genes)))
    }
  })
  
  observeEvent(input$copy_cart_genes, {
    genes <- cart_genes()
    
    if (length(genes) > 0) {
      showModal(modalDialog(
        title = "Copy Gene Cart",
        radioButtons("separator_choice", "Separator:",
                     choices = c("Comma" = ", ", "New Line" = "\n"),
                     selected = ", "),
        textAreaInput("gene_list_display", "Gene Cart (Copy with CTRL + A then CTRL + C)", "", rows = 15, cols = 100),
        footer = modalButton("Close")
      ))
      
      observe({
        separator <- input$separator_choice
        genes_string <- paste(genes, collapse = separator)
        updateTextAreaInput(session, "gene_list_display", value = genes_string)
      })
    } else {
      showNotification("Cart is empty.", type = "warning")
    }
  })
  
}

shinyApp(ui = ui, server = server)

```